### PR TITLE
ARROW-2499: [C++] Factor out Python iteration routines

### PR DESCRIPTION
**Pull request by pitrou**
Speeds up list to Arrow conversions by up to 15%. Also fixes a bug where creating a list array would not check that all input items are sequences.

Based on PR #1935.

### CODE REVIEW

**Comment by pitrou** [comment]
There are npm failures on Travis-CI (most probably unrelated, I'd say :-)).

File: cpp/src/arrow/python/builtin_convert.cc [hunk 1]
```diff
@@ -103,6 +113,12 @@ class ScalarVisitor {
         ss << type->ToString();
         return Status::Invalid(ss.str());
       }
+    } else if (PyList_Check(obj) || PyArray_Check(obj)) {
+      // XXX This code path is used for non-object arrays, which leads
```
Comment (xhochy) [inline cpp/src/arrow/python/builtin_convert.cc]: Can you make a JIRA for this and reference as `// TODO(ARROW-XXX): `?

Comment (pitrou) [reply cpp/src/arrow/python/builtin_convert.cc]: Done.

File: cpp/src/arrow/python/numpy-internal.h [hunk 2]
```diff
@@ -57,6 +57,7 @@ class Ndarray1DIndexer {
   T* begin() const { return data(); }
   T* end() const { return begin() + size() * stride_; }
 
+  // XXX should be `stride_ != 1`
   bool is_strided() const { return stride_ == 1; }
```
Comment (xhochy) [inline cpp/src/arrow/python/numpy-internal.h]: Is this is a problem that should be fixed here?

**Review by xhochy — COMMENTED** [review]
A bit of confused by one `XXX` otherwise this is ready to go

File: cpp/src/arrow/python/numpy-internal.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/python/numpy-internal.h]: I don't know. Perhaps the naming is simply wrong. I should take a look where it's used.

**Review by xhochy — APPROVED** [review]
+1, LGTM, merge on green build

**Comment by pitrou** [comment]
Looks like PR #1945 to solve the build issues.

**Comment by pitrou** [comment]
AppVeyor build: https://ci.appveyor.com/project/pitrou/arrow/build/1.0.366

**Pull request closed without merging**